pkg/base1/test-import-json.ts

```diff
@@ -0,0 +1,227 @@
+    assert.throws(() => import_string(12), /JSON validation error for : Not a string: 12/, "not a string");
+});
+
+QUnit.test("import_number", function(assert) {
+    assert.equal(import_number(12), 12, "number");
+    assert.throws(
+        () => import_number("foo"),
+        /JSON validation error for : Not a number: "foo"/,
+        "not a number");
+});
+
+QUnit.test("import_boolean", function(assert) {
+    assert.equal(import_boolean(true), true, "boolean");
+    assert.throws(
+        () => import_boolean("foo"),
+        /JSON validation error for : Not a boolean: "foo"/,
+        "not a boolean");
+});
+
+QUnit.test("import_array", function(assert) {
+    assert.deepEqual(import_array(["a", "b", "c"], import_string), ["a", "b", "c"], "array of strings");
+    assert.deepEqual(import_array([1, 2, 3], import_number), [1, 2, 3], "array of numbers");
+    assert.throws(
+        () => import_array([1, 2, "c"], import_number),
+        /JSON validation error for \[2\]: Not a number: "c"/,
+        "array of numbers with a string");
+    assert.throws(
+        () => import_array("foo", import_string),
+        /JSON validation error for : Not an array: "foo"/,
+        "not an array");
+});
+
+QUnit.test("import_record", function(assert) {
+    assert.deepEqual(
+        import_record({ a: "a", b: "b", c: "c" }, import_string),
+        { a: "a", b: "b", c: "c" },
+        "record of strings");
+    assert.deepEqual(
+        import_record({ a: 1, b: 2, c: 3 }, import_number),
+        { a: 1, b: 2, c: 3 },
+        "record of numbers");
+    assert.throws(
+        () => import_record({ a: 1, b: 2, c: "c" }, import_number),
+        /JSON validation error for \.c: Not a number: "c"/,
+        "record of numbers with a string");
+    assert.throws(
+        () => import_record("foo", import_string),
+        /JSON validation error for : Not an object: "foo"/,
+        "not a record");
+});
+
+QUnit.test("validate", function(assert) {
+    assert.equal(validate("test input", "foo", import_string, "default"), "foo", "string");
+    assert.equal(validate("test input", 12, import_string, "default"), "default", "not a string");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for test input: Not a string: 12'
+        ],
+        "console errors"
+    );
+});
+
+interface Player {
+    name: string;
+    age: number | undefined;
+}
+
+function import_Player(val: JsonValue): Player {
+    const obj = import_json_object(val);
+    return {
+        name: import_mandatory(obj, "name", import_string),
+        age: import_optional(obj, "age", import_number),
+    };
+}
+
+interface Team {
+    name: string;
+    players: Player[];
+}
+
+function import_Team(val: JsonValue): Team {
+    const obj = import_json_object(val);
+    return  {
+        name: import_mandatory(obj, "name", import_string),
+        players: import_mandatory(obj, "players", v => import_array(v, import_Player)),
```

Comment:
This looks pretty cool.

In general I kinda like how your way avoids adding `get_*()` for `str`, `number`, `obj`, `objv`, and a thousand other things...

Reply:
Now you get it. :-)